msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java

```diff
@@ -120,4 +80,75 @@ AuthenticationResult execute() throws Exception {
 
         return res;
     }
+
+    private AuthenticationResult makeRefreshRequest(AuthenticationResult cachedResult,  Authority requestAuthority) throws Exception {
+        RefreshTokenRequest refreshTokenRequest = new RefreshTokenRequest(
+                RefreshTokenParameters.builder(silentRequest.parameters().scopes(), cachedResult.refreshToken()).build(),
+                silentRequest.application(),
+                silentRequest.requestContext(),
+                silentRequest);
+
+        AcquireTokenByAuthorizationGrantSupplier acquireTokenByAuthorisationGrantSupplier =
+                new AcquireTokenByAuthorizationGrantSupplier(clientApplication, refreshTokenRequest, requestAuthority);
+
+        try {
+            AuthenticationResult refreshedResult = acquireTokenByAuthorisationGrantSupplier.execute();
+
+            refreshedResult.metadata().tokenSource(TokenSource.IDENTITY_PROVIDER);
+
+            log.info("Access token refreshed successfully.");
+            return refreshedResult;
+        } catch (MsalServiceException ex) {
+            //If the token refresh attempt threw a MsalServiceException but the refresh attempt was done
+            // only because of refreshOn, then simply return the existing cached token
+            if (!afterRefreshOn && (silentRequest.parameters().forceRefresh() || StringHelper.isBlank(cachedResult.accessToken()))) {
+                throw ex;
+            }
+            return cachedResult;
+        }
+    }
+
+    //Handles any logic to determine if a token should be refreshed, based on the request parameters and the status of cached tokens
+    private boolean shouldRefresh(SilentParameters parameters, AuthenticationResult cachedResult) {
+
+        //If forceRefresh is true, no reason to check any other option
+        if (parameters.forceRefresh()) {
+            setCacheTelemetry(CacheTelemetry.REFRESH_FORCE_REFRESH.telemetryValue);
+            return true;
+        }
+
+        //If the request contains claims then the token should be refreshed, to ensure that the returned token has the correct claims
+        //  Note: these are the types of claims found in (for example) a claims challenge, and do not include client capabilities
+        if (parameters.claims() != null) {
+            setCacheTelemetry(CacheTelemetry.REFRESH_FORCE_REFRESH.telemetryValue);
+            return true;
+        }
+
+        long currTimeStampSec = new Date().getTime() / 1000;
+
+        if (cachedResult.expiresOn() < currTimeStampSec) {
+            setCacheTelemetry(CacheTelemetry.REFRESH_ACCESS_TOKEN_EXPIRED.telemetryValue);
+            return true;
+        }
+
+        //Certain long-lived tokens will have a 'refresh on' time that indicates a refresh should be attempted long before the token would expire
+        if (cachedResult.refreshOn() != null && cachedResult.refreshOn() > 0 &&
+                cachedResult.refreshOn() < currTimeStampSec && cachedResult.expiresOn() >= currTimeStampSec){
```

Comment:
`cachedResult.expiresOn() >= currTimeStampSec` - not sure this is correct. MSALs use a 5 min buffer for this comparison.  Does MSAL Java not do this elsewhere? If not, please create a bug / consistency item to be fixed later (p2 at most). 5 min buffer is needed to account for time discrepancies.

Reply:
This logic was copied from where we set telemetry, which appears to have been missing the 5 minute buffer. However, [the method in TokenCache](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/fb028675c22d406ca6fdc9bcda6fa210d47fa806/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenCache.java#L483) that actually determines what tokens were pulled from the cache does correctly have the buffer.

In the latest commit I moved the logic for determining if a token is expired into AcquireTokenSilentSupplier to ensure the refresh logic is all in one place and telemetry is properly set.